Hidden powers of Bitcoin Blog (#53)
bitcoindevkit/bitcoindevkit.org: content/blog/2021/hidden_powers_of_bitcoin.md

---
title: "Miniscript Policy & Descriptors - Hidden Powers of Bitcoin"
description: "Introduction to Descriptor and Miniscript, making a Multisig Wallet and Testing Miniscript Policies"
authors:
    - Sandipan Dey
date: "2021-08-06"
tags: ["tutorial", "bdk", "bdk-cli", "miniscript", "descriptor", "bitcoin-cli"]
hidden: true
draft: false
---

To send people BTC - we simply scan a QR Code *(or paste an address)*, enter some amount and *whoosh* - sent!
Users might think, just like traditional currency, we can only exchange money using Bitcoin.
As it so happens, the underlying technology Bitcoin supports specify outputs not as addresses, but as programming scripts.
This opens us to a world of possibilities using Bitcoin.

### Script

Bitcoin supports [Script](https://en.bitcoin.it/wiki/Script), a **stack-based** lightweight programming language.

> Any script written in **Script** *(pun intended)* contains only `OP_*` codes that Bitcoin Full Nodes understand and process.

Reviewer: Script contains both `OP_*` codes and raw byte arrays.

> Currently, there are `186` op-codes in use.

Reviewer: `186` seems too much. As per my counting there are 117 opcodes (including OP_NOPs).  

Also maybe its better to link the wiki page for scripts instead the core code repo? Thats more descriptive and maybe helpful for new learners. https://en.bitcoin.it/wiki/Script

Author: Yes, this was a miscalculation on my part, `117` is correct 😅
And yup, the Wiki Page has better descriptions on each of the OP codes., adding it!

You can find a list of the op-codes straight from `bitcoin-core` code repository [here](https://github.com/bitcoin/bitcoin/blob/master/src/script/script.h#L65-L204).

Script is intentionally left [Turing incomplete](https://en.wikipedia.org/wiki/Turing_completeness) which is why there is no [halting problem](https://en.wikipedia.org/wiki/Halting_problem) with scripts.
There are no loops and overall, it's a very constrained programming language.

A transaction is considered valid only when an Output Script returns `true` at the end of execution.
Output Scripts define the conditions under which the coins associated with them can be spent.

> Really, to spend the output of a particular coin implies finding an input *(signature)* that makes a particular script evaluate to `true`.

Reviewer: Techincally "Output script" doesn't evaluate to true on its own. Below is a re-framing suggestion of the above paragraph. 

 ```suggestion
 A transaction is considered valid only when the Script returns `true` at the end of execution.
Output Script (aka scriptpubkey) define the conditions under which coins associated with them can be spent. To spend a particular coin implies finding an input script (aka scriptsig) such that a script made out of concatenation of `scriptsig + scriptpubkey` evaluates to `true`.
```

For example, a basic legacy `Pay-to-PubKey-Hash` transaction would look like:

```script
scriptPubKey: OP_DUP OP_HASH160 <pubKeyHash> OP_EQUALVERIFY OP_CHECKSIG
scriptSig: <sig> <pubKey>
```

##### Examples of things achievable using Bitcoin Script:

1. `Pay Someone (p2pkh/p2wpkh)` - A specific public key must sign to use the coins.
2. `Escrow (2-of-3-multisig)` - Two parties need to sign together to use the coins.
3. `Vault (locked)` - A specific key will not be able to use the coins for some time but another master key will always be able to use them.

> 4. `HTLC` - The receiver needs to ACK the tx before a time, else the coins are transferred back to the payee.

Reviewer: Suggestions inline. 
```suggestion
1. `Pay Someone (p2pkh/p2wpkh)` - A specific public key must sign to spend the coins.
2. `Escrow (2-of-3-multisig)` - Two parties need to sign together to spend the coins.
3. `Vault (locked)` - A specific key will not be able to spend the coins until a timeout but another master key will always be able to spend them.
4. `HTLC` - The receiver needs disclose a secret before a timeout, else the coins are transferred back to the payee.
```

##### Motivation for Policies

Unfortunately, due to its low-level and unusual stack-based nature, Script is pretty hard to reason about and use.

> Despite being around since Bitcoin creation, writing and understanding Script is not trivial.

Reviewer: ```suggestion
Despite being around since Bitcoin's creation, writing and understanding Script is not trivial.
```

This is why the scripts for the above few examples are pretty lengthy and might not make sense at the first glance.
When writing a script, we would want to know that if the logic we wrote is **correct**, **optimal** and **efficient in size** (use lower [weight](https://en.bitcoin.it/wiki/Weight_units)).

The community wanted an easy alternative way of writing Script that would create the most optimized Script code.
This gave rise to **Miniscript**.

### Miniscript

[Miniscript](http://bitcoin.sipa.be/miniscript/) tackles the above problems head-on.
It is an expressive way to create policies on Bitcoin Scripts in a structured and simple fashion.
Using Miniscript, it's difficult to go wrong.

Another very important goal of Miniscript is to replace any key used in a policy with another policy.
This is important because people might have multiple keys and complicated timelocks in their existing setup.
While signing a new policy, they would want to use this existing setup of theirs and also generate addresses for this setup.

> This is something accomplished using something called **Output Descriptors** which we will read about later.

Reviewer: This paragraph seemed a little confusing to me. Are you mentioning that we can replace a "key" or "value" inside a policy with another policy? Maybe I am wrong, but is this in general true?

In the last line you mentioned "Descriptors", so were you talking about replacing one "key" with another "key"? 

Nit:
  ```suggestion
Another very important goal of Miniscript is to replace any key used in a policy with another policy.
This is important because people might have multiple keys and complicated timelocks in their existing setup.
While signing a new policy, they would want to use their existing setup to also generate addresses for this new setup.
This is accomplished using something called **Output Descriptors** which we will get into in next section.
```

It doesn't contain any signature, it's mainly a combinator language for designing spending conditions.
You can try out the compiler online by using [this link](http://bitcoin.sipa.be/miniscript/#:~:text=Policy%20to%20Miniscript%20compiler).

##### Fragments

Here are some fragments which can be combined to create powerful expressions.

1. `pk(key)` - Specifies a given public key
2. `thresh(k, expr_1, expr_2, ..., expr_n)` - Specifies k of n multisig using expressions.
3. `older(T)` - Timelock for T blocks
4. `and(expr_1, expr_2)` - Both expressions should evaluate to true. 
5. `or(expr_1, expr_2)` - Any one of the expressions should evaluate to true.
6. `aor(expr_1, expr_2)` - Similar to `or` but `expr_1` has a more probability to evaluate to true.

Bitcoin Script allows us to use another alternate stack. The combinator functions use this second stack to evaluate expressions of `thresh`, `and`, `aor` and `or`.
The complete Miniscript Reference can be found [here](http://bitcoin.sipa.be/miniscript/#:~:text=Miniscript%20reference).

##### Example Policies

Here are the Miniscript Policies for the examples we looked at earlier. 

> Note `A`, `B`, `C` are placeholders for keys *(or output descriptors)* involved in the tx.

Reviewer: You cant put a descriptor inside miniscript, its other way around. 

Descriptors are high level description of scriptpubkey (p2pkh, p2sh etc). And miniscripts are semantics that describes the actual redeemscript. In general you have `Descriptor(Miniscript)` format.

1. Pay A (pay-to-public-key)
```
pk(A)
```

2. Escrow Account between A, B and third-party C.
```
thresh(2,pk(A),pk(B),pk(C))
```

3. Vault for A time-locked for T blocks with B as the master key.
```
aor(and(pk(A),time(T)),pk(B))
```

4. HTLC payment to B, which, if unspent for T blocks, returns to A.
```
aor(and(pk(A),time(T)),and(pk(B),hash(H))))
```

The Miniscript Policy Compiler is written in Rust and is present in [this repository](https://github.com/rust-bitcoin/rust-miniscript). 
In this blog, we will later use the same using [bitcoindevkit/bdk](https://github.com/bitcoindevkit/bdk), a lightweight descriptor-based wallet library
with a [cli](https://github.com/bitcoindevkit/bdk-cli). 

### Descriptors

Descriptors are a compact and efficient way to "describe" how scripts *(and addresses)* of a wallet should be generated.
They make it easier to deal with Multisig or complicated key setups.
Descriptors are super portable and can be easily used by any wallet to determine the list of all addresses that can be generated from the same.
This feature creates a common stage for all Bitcoin apps and software.

> It's an abstract way to replace keys.

Reviewer: To describe a little clearly the difference between a Descriptor and Miniscript, maybe you can add something along this line: 

"A bitcoin scriptpubkey can be of various types, ex P2PKH, P2SH, P2SH-P2WPKH etc. A descriptor is a simple description of what scriptpubkey scheme to be used for a given policy. A Descriptor can include a single pubkey within itself, or an entire miniscript policy. The miniscript policy will be used to derive the resulting redeemscript (the actual executable script), where as the descriptor describes how the redeemscript will be encumbered within the scriptpubkey."

According to Bitcoin Core, Output Descriptors are "a simple language which can be used to describe collections of output scripts".
They bring in themselves, derivation paths, master xpub/xprv fingerprints and paths to generate addresses from.
Let's understand this with an example of an Output Descriptor:

```

Descriptor: pkh([d34db33f/44'/0'/0']xpub6ERaJH[...]LJRcEL/1/*)#ml40v0wf
            <1> <--------2---------><----------3---------><4> <---5--->

Sections:
1 - function type (here, P2PK)

> 2 - derivation path of this key

Reviewer: Nit:
```suggestion
2 - master key fingerprint and derivation path from master
```

3 - xpub at m/44'/0'/0
4 - path to deriving keys/addresses at
5 - checksum for the descriptor
```

> This was an example of a simple descriptor using one function but they can also be used in conjugation with multiple functions like `combo`, `multi` etc.

Reviewer: `combo` and `multi` are two different things. `combo` is a descriptor which describes scriptpubkey of any type ex: P2PKH, P2SH etc.

`multi` is basically a miniscript which describes a m-of-n multisig script. When used as descriptor it will make the scriptpubkey as the raw multisig script. But its better to think it as a miniscript than a descriptor, because such raw multisigs are never used in practice.

In wild you will see stuffs like `wsh(multi(2, PKA, PKB, PKC))`. The `wsh` part is the descriptor describing its a P2WSH output, and `multi` part is the miniscript describing the actual spending condition.

Ya it gets messy with terminology in bitcoin scripts. But if we aren't careful, this can cause more confusion to readers.

Author: True, I added this section in the blog because it makes the concept clearer.

> The concept of descriptor came into existence in 2018 and since then, a lot of wallets have added support for descriptors.

Reviewer: Might be a good idea to include a link for descriptor doc here somewhere. https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md

During this time, we will try to break our vault and see our transactions failing.

##### Tools and Armor

- [docker](https://docs.docker.com/engine/install/)
- [bdk-cli](https://github.com/bitcoindevkit/bdk-cli)
- [miniscriptc](https://bitcoindevkit.org/bdk-cli/compiler/#installation)

##### Setting Up

We require `bitcoind` to run in `regtest` mode.
The hard work has been already done in the `bitcoindevkit/electrs` docker image.
The electrs functionality will also prove significant because `bdk` can use this local `electrs` box and send the transactions that we're gonna do for testing.

```bash
alias elstart='docker run --detach --rm -p 127.0.0.1:18443-18444:18443-18444/tcp -p 127.0.0.1:60401:60401/tcp --name electrs bitcoindevkit/electrs:0.5.0'
alias elstop='docker kill electrs'
alias ellogs='docker container logs electrs'
alias elcli='docker exec -it electrs /root/bitcoin-cli -regtest -datadir=/root/.bitcoin $@'

# Start Bitcoin Core
elstart

# Delete default wallet created by bitcoindevkit/electrs
elcli unloadwallet "bdk-test"
```

Please note that `elcli` is an alias for `bitcoin-cli`.

#### Keys and Generating Addresses

> Let us first generate an XPRV and create the wpkh wallet descriptor

Reviewer: Maybe before this step its better to include installation instruction for bdk-cli? 

bdk-cli would also needs a backend to work with. In latest update the default electrum backend is removed. So in order to use bdk-cli for this tutorial they need to compile it with `--features electrum --features compiler`.

Author: I tried `cargo install bdk-cli --features electrum --features compiler` but this didn't complete successfully.
It's ending in: 
```
error: no binaries are available for install using the selected features
```

Reviewer: Ah that's because the new version 0.2.1 of bdk-cli hasn't been tagged yet. cc @notmandatory . Currently cargo is trying to install 0.2.0, which requires the `repl` feature and is not included by default.

This works with 0.2.0 `cargo install bdk-cli --features electrum --features compiler --features repl`.

But we should get the latest bdk-cli tagged before releasing this tutorial. Otherwise the instructions can break in future.

Reviewer: Also curious, what version of bdk-cli did you use in the tutorial?

bcrt1q4p4k63xglftez0h8yc7d4kmhsn5j5kecguu34j
bcrt1q29z2uanskweur7qrzr43gyv3l028s0pnd9ptvp
bcrt1qkzpeqz8sd73sucfythjxftez0e3ee30yhp9w67
bcrt1qptwd6ggy8ttryck2f6yjf4la68apruc3fs7elz

# Generation via bitcoin-cli
❯ repeat 10 { elcli -rpcwallet="mywallet" getnewaddress }
bcrt1qc9wzxf8pthyexl00m23ug92pqrthagnzzf33wp
bcrt1qgnh7e72q92fqujwg3qxlg5kplxkm6rep0nerur
bcrt1qea6r8yvd0peupk29p94wm0xasvydgdsnyzkhez
bcrt1qm99230tpqflq0f8kpkn5d2tee02hgqcsw5sd99
bcrt1qd0afjfnl5udrsfkrj72rl34pss34yluma752qv
bcrt1qj2aymplrzxcp4m7vcxrzq93g58pmgm4fpluesy
bcrt1q4p4k63xglftez0h8yc7d4kmhsn5j5kecguu34j
bcrt1q29z2uanskweur7qrzr43gyv3l028s0pnd9ptvp
bcrt1qkzpeqz8sd73sucfythjxftez0e3ee30yhp9w67
bcrt1qptwd6ggy8ttryck2f6yjf4la68apruc3fs7elz
```

Notes:
- The `repeat n {}` syntax will only work in `zsh`
- In case you get different outputs in either of the cases, try deleting `~/.bdk-bitcoin` and retrying (thanks [@Steve](https://twitter.com/notmandatory) for this tip!)

So now we have definitive proof that descriptors can make wallets portable. That single string will be able to make any wallet generate the same set of addresses and hence they can sync and broadcast transactions in the same manner!

> #### Making a Secure Descriptor for Funds

Reviewer: It seems this a "Multisig" transaction section. "Secure" is generic term and can mean many things.

{
  "satoshi": 2500000000
}

❯ bdk-cli -n regtest wallet --server tcp://127.0.0.1:60401 -w K3 -d $K3_DESC sync
{}
❯ bdk-cli -n regtest wallet --server tcp://127.0.0.1:60401 -w K3 -d $K3_DESC get_balance
{
  "satoshi": 2500000000
}
```

Everyone has the same amount of balance. 
This happened because it was a multisig wallet.
Now, let's try to spend some balance.
We will give back some balance to the wallet maintained by `bitcoin-cli`.
But remember, this is a `2-of-3` multisig wallet.
That's why we will need atleast two keys to sign to make a transaction.

Here's where we will require to use a [PSBT](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki) or a *partially signed bitcoin transaction*. 
Bitcoin uses PSBTs as the standard protocol to create a transaction and add one or more signatures to it before broadcasting the same to 
the network which finally can become a proper valid *transaction*.

```bash
# address to send tx to from this multisig wallet
elcli getnewaddress

# create the transaction, can be started by anyone
PSBT=$(bdk-cli -n regtest wallet --server tcp://127.0.0.1:60401 -w K2 -d $K2_DESC \

> 	create_tx --to "<address-from-last-step>:5555555" | jq -r ".psbt")

Reviewer: Instead of sending address from last step its better to store it in a variable. Its done like that everywhere else.

bdk-cli -n regtest wallet --server tcp://127.0.0.1:60401 -w K3 -d $K3_DESC sync
bdk-cli -n regtest wallet --server tcp://127.0.0.1:60401 -w K3 -d $K3_DESC get_balance
{
  "satoshi": 2494444255
}
```

So this proves we can definitely do transactions with multisig wallets with complicated descriptors.
Since for Bitcoin, having keys equal having access to the accounts, we need to keep our keys safe.
For legacy single key wallets, we used to keep backups of the mnemonic codes in multiple places.
It was pretty insecure because in case any one of those backups leaks, our entire account would be compromised.
Complicated multisig wallet descriptors are definitely a step forward - just in case a single key leak or are lost, no one would be able to take charge of the funds we hold.

### Retention Bonus - Smart Contract with Bitcoin

Let us consider that a company wants to give its employees a retention bonus for two months.
If an employee stays with that company for over 2 months, the employee would get 1 BTC as a reward.
This would be a smart contract between the company and an employee.
The employee should be able to see that he would get his funds after a year.

> The company would require confidence that the employee would not be able to withdraw the reward before 1 year has passed.

Reviewer: Shouldn't the employee get the funds after 2 months (not 1 year)?

The Miniscript policy for this contract would be as follows:
```

> or(99@and(pk(E),older(8640)),pk(C))

Reviewer: Either I am reading it wrong or 8640 is blocks worth of 1 month, not 2.

Author: Block Time = 10 mins
2 Months = 2*30*24*60 mins = 86400 mins
n(Blocks) = 8640

Am I doing it wrong somewhere? 😅

Reviewer: Ah my bad.. I made stupid mistake in my calculation. This is correct.

```
where `E` is the employee and `C` is the company.

I should emphasize over here that this policy will let the company still transfer funds after the designated 2 months.
It's not possible to block them after the lock time has passed, atleast not in a single policy.

Surely, after two months, the funds can be unlocked by the employee but before that, the company can revoke the funds.

> Let us compile this policy down to miniscript.

Reviewer: The policy is compiled down to a `descriptor`. Not another miniscript.

```bash
# The Descriptor will be on the log, the E and C are placeholders

> ❯ miniscriptc "or(99@and(pk(E),older(8640)),pk(C))" sh-wsh

Reviewer: This probably old version of bdk-cli. Currently it doesn't produce `miniscriptc` anymore. we have to use `bdk-cli compile` command.

Reviewer: Also please confirm this with `v0.2.0` as far as I can see there is no `miniscriptc` binary anymore.

[2021-08-05T12:25:40Z INFO  miniscriptc] Compiling policy: or(99@and(pk(E),older(8640)),pk(C))
[2021-08-05T12:25:40Z INFO  miniscriptc] ... Descriptor: sh(wsh(andor(pk(E),older(8640),pk(C))))#55wzucxa
Error: Descriptor(Miniscript(Unexpected("Key too short (<66 char), doesn't match any format")))
```

So the compiled miniscript is
```

> sh(wsh(andor(pk(E),older(8640),pk(C))))

Reviewer: This is a descriptor. The miniscript is inside the descriptor.
